Northstar.CustomServers/mod/scripts/vscripts/mp/_stats.nut

```diff
@@ -12,67 +14,1083 @@ global function PreScoreEventUpdateStats
+			
+			if ( state != "" )
+				Stats_IncrementStat( player, "time_stats", state, "", timeHours )
+
+			// weapon time stats
+			entity activeWeapon = player.GetActiveWeapon()
+			if ( IsValid( activeWeapon ) )
+			{
+				if ( IsValidStatItemString( activeWeapon.GetWeaponClassName() ) )
+					Stats_IncrementStat( player, "weapon_stats", "hoursUsed", activeWeapon.GetWeaponClassName(), timeHours )
+
+				foreach( entity weapon in player.GetMainWeapons() )
+				{
+					if ( IsValidStatItemString( weapon.GetWeaponClassName() ) )
+						Stats_IncrementStat( player, "weapon_stats", "hoursEquipped", weapon.GetWeaponClassName(), timeHours )
+				}
+			}
+
+			// map time stats
+			Stats_IncrementStat( player, "game_stats", "hoursPlayed", "", timeHours )
+		}
+
+		lastTickTime = Time()
+		// not rly worth doing this every frame, just a couple of times per second should be fine
+		wait 0.25
+	}
+}
+
+// this is kinda shit
+void function SaveStatsPeriodically_Threaded()
+{
+	while( true )
+	{
+		foreach( entity player in GetPlayerArray() )
+			Stats_SaveAllStats( player )
+		wait 5
+	}
 }
 
-void function PreScoreEventUpdateStats(entity attacker, entity ent)
+bool function IsValidGamemodeString( string mode )
 {
+	int gameModeCount = PersistenceGetEnumCount( "gameModes" )
+	for ( int modeIndex = 0; modeIndex < gameModeCount; modeIndex++ )
+	{
+		string gameModeName = PersistenceGetEnumItemNameForIndex( "gameModes", modeIndex )
 
+		if ( gameModeName == mode )
+			return true
+	}
+
+	return false
 }
 
-void function PostScoreEventUpdateStats(entity attacker, entity ent)
+bool function IsValidStatItemString( string item )
 {
+	foreach( str in shGlobalMP.statsItemsList )
+	{
+		if ( str == item )
+			return true
+	}
 
+	return false
 }
 
-void function Stats_OnPlayerDidDamage(entity player, var damageInfo)
+string function GetPersistenceRefFromDamageInfo( var damageInfo )
 {
+	string damageSourceString = DamageSourceIDToString( DamageInfo_GetDamageSourceIdentifier( damageInfo ) )
+
+	foreach( str in shGlobalMP.statsItemsList )
+	{
+		if ( str == damageSourceString )
+			return damageSourceString
+	}
 
+	return ""
 }
+
+bool function DamageIsTitanMelee( int damageSourceId )
+{
+	if ( damageSourceId == eDamageSourceId.melee_titan_punch )
+		return true
+	else if ( damageSourceId == eDamageSourceId.melee_titan_punch_ion )
+		return true
+	else if ( damageSourceId == eDamageSourceId.melee_titan_punch_legion )
+		return true
+	else if ( damageSourceId == eDamageSourceId.melee_titan_punch_tone )
+		return true
+	else if ( damageSourceId == eDamageSourceId.melee_titan_punch_scorch )
+		return true
+	else if ( damageSourceId == eDamageSourceId.melee_titan_punch_northstar )
+		return true
+	else if ( damageSourceId == eDamageSourceId.melee_titan_punch_fighter )
+		return true
+	else if ( damageSourceId == eDamageSourceId.melee_titan_sword )
+		return true
+	else if ( damageSourceId == eDamageSourceId.melee_titan_sword_aoe )
+		return true
+
+	return false
```

Comment:
```suggestion
	switch( damageSourceId )
	{
		case eDamageSourceId.melee_titan_punch:
		case eDamageSourceId.melee_titan_punch_ion:
		case eDamageSourceId.melee_titan_punch_legion:
		case eDamageSourceId.melee_titan_punch_tone:
		case eDamageSourceId.melee_titan_punch_scorch:
		case eDamageSourceId.melee_titan_punch_northstar:
		case eDamageSourceId.melee_titan_punch_fighter:
		case eDamageSourceId.melee_titan_sword:
		case eDamageSourceId.melee_titan_sword_aoe:
			return true
		default:
			return true
	}
	unreachable
```

Reply:
This was a copied function from s2s, but yeah this is much better than whatever respawn made